include/envoy/server/listener_manager.h

```diff
@@ -23,11 +23,13 @@ class ListenerComponentFactory {
   /**
    * Creates a socket.
    * @param address supplies the socket's address.
+   * @param transparent supplies whether to configure the socket as transparent.
    * @param bind_to_port supplies whether to actually bind the socket.
    * @return Network::SocketSharedPtr an initialized and potentially bound socket.
    */
   virtual Network::SocketSharedPtr
-  createListenSocket(Network::Address::InstanceConstSharedPtr address, bool bind_to_port) PURE;
+  createListenSocket(Network::Address::InstanceConstSharedPtr address, bool transparent,
```

Comment:
Q: Do we foresee having more options like this in the future? If so should we switch to some type of options struct that gets passed around? That would a avoid having to change every individual callsite and test again in the future potentially. Just throwing it out there.

Reply:
I personally don't foresee more options like this. But if we think that could happen, I could use https://github.com/envoyproxy/envoy/pull/2734 to implement this internally.

Reply:
If we don't foresee more options like this than I wouldn't worry about it.

Reply:
I have one more option coming that is similar/related: enabling tcp-fastopen on listening sockets.

Reply:
@ggreenway So what is your preference? Should we implement this with #2734?

Reply:
I haven't read #2734 carefully, but it looks like there can only be one option-setter, so if both a ListenerFilter wants to set an option and there is an option from the config, we'd need some more code to combine the two, or somehow not lose one of them. So I'm unsure if that's the best way to do it or not.

@jrajahalme do you have an opinion on this?

Reply:
I was under the impression we already had an option to set IP_FREEBIND on listening sockets, but it looks like this isn't true. Expect a PR for that one to come very soon.

Reply:
OK given several options that we need, can we do the work to make an options struct? This will avoid having to change a ton of callsites every time we add a new option.

Reply:
I'm planning to begin work on the IP_FREEBIND implementation later today. If we're ready to merge this in it's current form I'm happy to struct-ify the options as part of that PR so @rlenglet doesn't have to make the change here.

Reply:
question.. how does IP_FREEBIND differ from IP_TRANSPARENT?

Reply:
@ggreenway Would it work for you better if I generalized the current option setting code (i.e., #2734 ) to accept multiple option setters, maintain a list of set options and have the call sites call all of them (unless one of them fails)?

Reply:
@dnoe OK SGTM if you want to do the struct stuff.

Reply:
@jrajahalme yes that sounds reasonable

Reply:
FREEBIND allows binding to an address that isn't configured on any interface on the machine. IP_TRANSPARENT would also allow this, but from inside Envoy should otherwise work the same as a regular listener (whereas transparent is designed around a different use case where the traffic has been redirected to Envoy and has some original destination that is meaningful).